### PR TITLE
simplify max_size and friends

### LINKED ISSUE #1781: simplify max_size and friends
## Description
max_size, max_size_mvt, check_consistent_sizes, and check_consistent_sizes_mvt all hardcode the 2-3-4-5 arguments cases. They can be simplified by rewriting them as folds over parameter packs, and by proper use of std::max.

This has been brought up a few times, but no one ever made an issue for it.

## Example
std::max can be used like this:
```
template <typename T1, typename... Ts>
size_t max_size(const T1& x1, const Ts&... xs) {
  return std::max({stan::math::size(x1), stan::math::size(xs)...});
}
```

A fold over a parameter pack looks like this:
```
int sum(int accum) {
  return accum;
}

template <typename T1, typename... Ts>
int sum(int accum, T1 x1, Ts... xs) {
  return sum(accum + x1, xs...);
}
```

Actually in C++17 that can be just:
```
template <typename T1, typename... Ts>
int sum(int accum, T1 x1, Ts... xs) {
  return (x1 + ... + xs);
}
```
But C++17's fold expressions only work for a select group of blessed operators, you can't just provide a binary function to fold with. So even if we were on 17 we couldn't take advantage of that.

## Expected Output
The same as before.

#### Current Version:
v3.1.0

### PR DESCRIPTION
**Pull request by peterwicksstringfield**
##  Summary

Reimplement max_size, max_size_mvt, check_consistent_sizes, check_consistent_sizes_mvt in a simpler and more general way.

All these functions can now be called with any natural number of arguments. The 1 argument case is sort of trivial but I see no harm in allowing it.

Closes #1781.

## Tests

I added a few tests for all 4 functions.

runTests.py test/unit/math/prim/err/check_consistent_sizes_mvt_test.cpp test/unit/math/prim/fun/max_size_mvt_test.cpp test/unit/math/prim/fun/max_size_test.cpp test/unit/math/rev/err/check_consistent_sizes_test.cpp

## Side Effects

Shouldn't be any.

## Release Notes

max_size, max_size_mvt, check_consistent_sizes, and check_consistent_sizes_mvt have been simplified and now take any number of arguments.

## Checklist

- [X] Math issue #1781

- [X] Copyright holder: Peter Wicks Stringfield

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [X] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [X] the code is written in idiomatic C++ and changes are documented in the doxygen

- [X] the new changes are tested

### CODE REVIEW

**Comment by bob-carpenter** [comment]
I can review this later this afternoon.

**Comment by peterwicksstringfield** [comment]
I can probably anticipate some of your comments. Maybe hold off on review for a second?

**Comment by peterwicksstringfield** [comment]
There. Once that finishes building this will be ready for review.

File: stan/math/prim/err/check_consistent_sizes.hpp [hunk 1]
```diff
@@ -1,150 +1,84 @@
...
-                            is_vector<T5>::value * stan::math::size(x5)))));
-  check_consistent_size(function, name1, x1, max_size);
-  check_consistent_size(function, name2, x2, max_size);
-  check_consistent_size(function, name3, x3, max_size);
-  check_consistent_size(function, name4, x4, max_size);
-  check_consistent_size(function, name5, x5, max_size);
+                                   const T1& x1, const Ts&... names_and_xs) {
+  size_t max_size
+      = std::max({internal::size_ignore_nonvector(x1),
+                  internal::size_ignore_nonvector(names_and_xs)...});
```
Comment (bob-carpenter) [inline stan/math/prim/err/check_consistent_sizes.hpp]: [comment]
This idiom is super cool.  I didn't realize you could use a brace initializer that way without putting everything in a tuple.

File: stan/math/prim/err/check_consistent_sizes_mvt.hpp [hunk 2]
```diff
@@ -3,31 +3,83 @@
...
 
 namespace stan {
 namespace math {
+namespace internal {
+
+/** Base case of recursion, this function is a no-op. */
+inline void check_consistent_with_size_mvt(const char*, size_t) { return; }
+
+/**
+ * Check that the provided inputs are either non-vectors, or vectors of the
```
Comment (bob-carpenter) [inline stan/math/prim/err/check_consistent_sizes_mvt.hpp]: This doc is identical to that for `check_consist_with_size`.  This doc shold explain what the `mvt` means in the suffix.

File: stan/math/prim/fun/max_size.hpp [hunk 3]
```diff
@@ -2,43 +2,21 @@
...
+/** Calculate the size of the largest input.
+ * @tparam T1 type of the first input
+ * @tparam Ts types of the other inputs
+ * @param x1 first input
+ * @param xs other inputs
+ * @return the size of the largest input
+ */
+template <typename T1, typename... Ts>
+inline size_t max_size(const T1& x1, const Ts&... xs) {
+  return std::max({stan::math::size(x1), stan::math::size(xs)...});
```
Comment (bob-carpenter) [inline stan/math/prim/fun/max_size.hpp]: Can this be generalized to handle size zero inputs as well as the existing generalization to one element?  Though I don't know what max returns with no inputs---negative infinity?  

```
template <typename... Ts>
 inline size_t max_size(const Ts&... xs) {
   return std::max({stan::math::size(xs)...});
```

File: stan/math/prim/fun/max_size.hpp [hunk 4]
```diff
@@ -2,43 +2,21 @@
...
-template <typename T1, typename T2, typename T3, typename T4, typename T5>
-size_t max_size(const T1& x1, const T2& x2, const T3& x3, const T4& x4,
-                const T5& x5) {
-  size_t result = stan::math::size(x1);
-  result = result > stan::math::size(x2) ? result : stan::math::size(x2);
-  result = result > stan::math::size(x3) ? result : stan::math::size(x3);
-  result = result > stan::math::size(x4) ? result : stan::math::size(x4);
-  result = result > stan::math::size(x5) ? result : stan::math::size(x5);
-  return result;
+/** Calculate the size of the largest input.
```
Comment (bob-carpenter) [inline stan/math/prim/fun/max_size.hpp]: The style used elsewhere is to have just a line with `/**` to start a comment.

File: stan/math/prim/err/check_consistent_sizes.hpp (hunk 1 above)
Comment (peterwicksstringfield) [reply stan/math/prim/err/check_consistent_sizes.hpp]: This ones new to me as well.

File: stan/math/prim/fun/max_size.hpp (hunk 3 above)
Comment (peterwicksstringfield) [reply stan/math/prim/fun/max_size.hpp]: Yes. negative infinity isn't a size_t, but 0 is. I think where this comes up is if someone does max_size(xs...) and the parameter pack xs turns out to be empty, or maybe at the bottom of a meta programming recursion. I wonder what they would expect max_size([empty]) to be. A compiler error is a safer surprise than an unexpected 0, but 0 does seem like something someone might expect. Certainly no one would ever write deliberately write "max_size()".

File: stan/math/prim/fun/max_size.hpp (hunk 4 above)
Comment (peterwicksstringfield) [reply stan/math/prim/fun/max_size.hpp]: Thanks. Fixed.

File: stan/math/prim/err/check_consistent_sizes_mvt.hpp (hunk 2 above)
Comment (peterwicksstringfield) [reply stan/math/prim/err/check_consistent_sizes_mvt.hpp]: Whoops. Meant to fix that. Thanks!

Comment (peterwicksstringfield) [reply stan/math/prim/err/check_consistent_sizes_mvt.hpp]: I'm also cleaning up check_consistent_size_mvt (Note: "size" not "sizes"). The logic and documentation there is pretty confusing and can be simplified.

**Comment by syclik** [comment]
@bob-carpenter, mind taking a look at your review to see if everything's been addressed?

**Pull request merged**